test/hooks_test.lua

```diff
@@ -62,6 +62,60 @@ g.test_hooks = function()
     t.assert_equals(hooks, expected)
 end
 
+g.test_custom_hooks = function()
+    local _hooks = require('luatest.hooks')
+    local hooks = {}
+
+    _hooks.before_suite_preloaded(function() table.insert(hooks, 'before_suite') end)
+    _hooks.after_suite_preloaded(function() table.insert(hooks, 'after_suite') end)
+    _hooks.before_suite_preloaded(function() table.insert(hooks, 'before_suite2') end)
+    _hooks.after_suite_preloaded(function() table.insert(hooks, 'after_suite2') end)
+
+    _hooks.before_all_preloaded(function() table.insert(hooks, 'before_all') end)
+    _hooks.after_all_preloaded(function() table.insert(hooks, 'after_all') end)
+    _hooks.before_all_preloaded(function() table.insert(hooks, 'before_all2') end)
+    _hooks.after_all_preloaded(function() table.insert(hooks, 'after_all2') end)
+
+    _hooks.before_suite_preloaded(function() table.insert(hooks, 'before_suite3') end)
+    _hooks.before_all_preloaded(function() table.insert(hooks, 'before_all3') end)
+    _hooks.before_all_preloaded(function() table.insert(hooks, 'before_all4') end)
+    _hooks.after_suite_preloaded(function() table.insert(hooks, 'after_suite3') end)
+    _hooks.after_all_preloaded(function() table.insert(hooks, 'after_all3') end)
+    _hooks.after_all_preloaded(function() table.insert(hooks, 'after_all4') end)
+
+    local result = helper.run_suite(function(lu2)
+        local t2 = lu2.group('test')
+        t2.before_all(function() table.insert(hooks, 'before_all_inner') end)
+        lu2.before_suite(function() table.insert(hooks, 'before_suite_inner') end)
+        lu2.after_suite(function() table.insert(hooks, 'after_suite_inner') end)
+        t2.after_all(function() table.insert(hooks, 'after_all_inner') end)
+        t2.test = function() table.insert(hooks, 'test') end
+    end)
+
+    t.assert_equals(result, 0)
+    t.assert_equals(hooks, {
+        "before_suite",
+        "before_suite2",
+        "before_suite3",
+        "before_suite_inner",
+        "before_all",
+        "before_all2",
+        "before_all3",
+        "before_all4",
+        "before_all_inner",
+        "test",
+        "after_all4",
+        "after_all3",
+        "after_all2",
+        "after_all",
+        "after_all_inner",
```

Comment:
I would appreciate if the 'inner' after\* hooks run before the predefined hooks.

Reply:
It seems, we can adjust the `run_` functions to implement this order.

```diff
diff --git a/luatest/hooks.lua b/luatest/hooks.lua
index 8708543..fc2c989 100644
--- a/luatest/hooks.lua
+++ b/luatest/hooks.lua
@@ -17,7 +17,7 @@ function export.before_suite_preloaded(fn)
 end
 
 function export.after_suite_preloaded(fn)
-    table.insert(preloaded_hooks.after_suite, 1, {fn, {}})
+    table.insert(preloaded_hooks.after_suite, {fn, {}})
 end
 
 function export.before_all_preloaded(fn)
@@ -25,7 +25,7 @@ function export.before_all_preloaded(fn)
 end
 
 function export.after_all_preloaded(fn)
-    table.insert(preloaded_hooks.after_all, 1, {fn, {}})
+    table.insert(preloaded_hooks.after_all, {fn, {}})
 end
 
 local function check_params(required, actual)
@@ -59,17 +59,45 @@ local function define_hooks(object, hooks_type, preloaded_hook)
     end
     object['_original_' .. hooks_type] = object[hooks_type] -- for leagacy hooks support
 
+    local function run_preloaded_hooks()
+        if preloaded_hook == nil then
+            return
+        end
+
+        -- before_* -- direct order
+        -- after_* -- reverse order
+        local from = 1
+        local to = #preloaded_hook
+        local step = 1
+        if hooks_type:startswith('after_') then
+            from, to = to, from
+            step = -step
+        end
+
+        for i = from, to, step do
+            local hook = preloaded_hook[i]
+            if check_params(hook[2], object.params) then
+                hook[1](object)
+            end
+        end
+    end
+
     object['run_' .. hooks_type] = function()
+        -- before_* -- run before test hooks
+        if hooks_type:startswith('before_') then
+            run_preloaded_hooks()
+        end
+
         local active_hooks = object[hooks_type .. '_hooks']
         for _, hook in ipairs(active_hooks) do
             if check_params(hook[2], object.params) then
                 hook[1](object)
             end
         end
-    end
-    if preloaded_hook then
-        for _, v in ipairs(preloaded_hook) do
-            table.insert(hooks, v)
+
+        -- after_* -- run after test hooks
+        if hooks_type:startswith('after_') then
+            run_preloaded_hooks()
         end
     end
 end
diff --git a/test/hooks_test.lua b/test/hooks_test.lua
index d9e7478..cec69a7 100644
--- a/test/hooks_test.lua
+++ b/test/hooks_test.lua
@@ -104,15 +104,15 @@ g.test_custom_hooks = function()
         "before_all4",
         "before_all_inner",
         "test",
+        "after_all_inner",
         "after_all4",
         "after_all3",
         "after_all2",
         "after_all",
-        "after_all_inner",
+        "after_suite_inner",
         "after_suite3",
         "after_suite2",
         "after_suite",
-        "after_suite_inner",
     })
 end
```

Reply:
Added.